### ISSUE TITLE
getAllMarkers() (ProblemsView.ts) isn't reliable

### ISSUE THREAD

**Issue by jeffb-sfdc** [redhat-developer/vscode-extension-tester#487]
TL;DR: I found an issue with `getAllMarkers()` (in `page-objects/src/components/bottomBar/ProblemsView.ts`), and am posting  a PSA to warn other developers.  In the Problems view, there may be "N" issues reported, but the view lazy-loads the rows, so using CSS selectors to get the count **only gets the count of the items which are currently visible, and not the actual count**.  This is an issue with VS Code, but when writing e2e tests, getAllMarkers() should not be used.

The full details:
I'm new to vscode-extension-tester and current evaluating it.  I'm going through the tests in `helloworld-sample` to get a feel for how to use vscode-extension-tester.  I was expecting 100% of the tests to pass, but so far I've encountered many issues.  In bottomBar-test.ts, there are two tests: "Error markers are displayed" and "There are no warnings".  I found these to be flakey (and in the case of "There are no warnings", flat out wrong), so I changed them to:
```
        // now we can look at the error markers
        it('Error markers are displayed', async () => {
            // generally, there are 3 marker types (warning, error, and file - file just contains other markers though)
            // we want to see the errors
            const errors = await view.getAllMarkers(MarkerType.Error);

            // assert that there are errors (there should be about 8 in the file)
            // expect(errors.length).is.greaterThan(5);
            expect(errors.length).equals(7); // look for the exact count
        });

        // now we can look at the warning markers
        it('Warning markers are displayed', async () => {
            const warnings = await view.getAllMarkers(MarkerType.Warning);

            expect(warnings.length).equals(12);
        });
```

When I run the suit of tests in bottomBar-test.ts, and problems.ts is loaded into the instance of VS Code that's being tested, there are 7 errors and 12 warnings.

When I run on a large (secondary) monitor, and all items in the list are visible, the two tests I just posted pass.

On the other hand, when the tests (and VS Code) run on normal-sized laptop monitor, not all items in the Problems view are visible, and the tests fail.  This is because VS Code is lazy loading (and unloading) items in this list.

I looked into the implementation of `getAllMarkers()` (in page-objects/src/components/bottomBar/ProblemsView.ts) and `findElements()` is getting called with `ProblemsView.locators.ProblemsView.markerRow`:
```
    async getAllMarkers(type: MarkerType): Promise<Marker[]> {
        const markers: Marker[] = [];
        const elements = await this.findElements(ProblemsView.locators.ProblemsView.markerRow);
        .
        .
        .
```
The CSS selector for markerRow is '.monaco-list-row'.  I opened the Developer Tools in VS Code and was able to verify that the rows have a CSS class of '.monaco-list-row', and that as I expand/contract the height of the Problems view, the count changes:
```
document.getElementsByClassName('monaco-list-row').length
```

Not sure what to do here about this... I'm sure Microsoft is going to say that this is by design... after all, if the view had 1,000 issues listed, loading 1,000 rows into memory but only displaying a fraction, like 50, is a waste.  On the other hand, getAllMarkers() depends on the state of the UI and how many items are currently displayed, and isn't reliable.

**Comment by mlorinc** [comment]
Hi @jeffb-sfdc,

thank you for your feedback. I agree `getAllMarkers` method has misleading name which does not match its implementation. The question is if it is practical to load all markers at once. It is possible to store marker data (label, type, position, etc.) for later use but it is not possible to interact with any marker which is unloaded from DOM. If we stored marker positions in scroll container any UI update or event could invalidate marker position in any time.

In my opinion it should be handled the same we handle quickpicks in [Input.ts](https://github.com/redhat-developer/vscode-extension-tester/blob/master/page-objects/src/components/workbench/input/Input.ts). 

In this case it would bring the following changes:

1. `getAllMarkers` is set to deprecated
2. new getter methods: `getMarkers` or `getVisibleMarkers`
3. findMarker method which would accept predicate as argument -- `(marker) => PromiseLike<boolean> | boolean`
4. new methods for scrolling: `nextPage`, `previousPage`, `hasNextPage` and `hasPreviousPage`

What do you think about this proposal @jeffb-sfdc, @odockal and @djelinek?

**Comment by jeffb-sfdc** [comment]
@mlorinc I would suggest the following:
1. rename `getAllMarkers()` to `getAllVisibleMarkers()`
2. change src/ui-test/bottomBar-test.ts to
```
        // These tests use getAllMarkers() and are unreliable and should not be included.
        //
        // now we can look at the error markers
        // it('Error markers are displayed', async () => {
        //     // generally, there are 3 marker types (warning, error, and file - file just contains other markers though)
        //     // we want to see the errors
        //     const errors = await view.getAllMarkers(MarkerType.Error);
        //
        //     // assert that there are errors (there should be about 8 in the file)
        //     expect(errors.length).is.greaterThan(5);
        // });
        //
        // // we can make sure no warnings are present at the same time
        // it('There are no warnings', async () => {
        //     const warnings = await view.getAllMarkers(MarkerType.Warning);
        //     expect(warnings).is.empty;
        // });
        //
        // // there is also a file marker (out problematic file that contains the errors)
        // it('There is a file marker', async () => {
        //     const files = await view.getAllMarkers(MarkerType.File);
        //     const file = files[0];
        //
        //     // we can get the text of the marker
        //     expect(await file.getText()).contains('problems.ts');
        //     // and the type
        //     expect(await file.getType()).equals(MarkerType.File);
        //     // and we can collapse & expand the file marker
        //     await file.toggleExpand(false);
        //     await file.toggleExpand(true);
        // });

        it('Markers are displayed', async () => {
            // Need to throttle this test in order for VS Code to load/display all of the errors
            // and warnings.
            await new Promise(res => setTimeout((res), 3000));

            const markers = await view.getAllMarkers(MarkerType.Any);
            const badgeElement = await view.findElement(By.css('.monaco-count-badge'));
            const badgeText = await badgeElement.getText();

            // getAllMarkers() only returns the **visible** markers, so we can't rely on the count,
            // but we should be able to rely on at least one appearing.
            expect(markers.length).is.greaterThan(0);
            expect(badgeText).equals("19");
        });
```

I'm going to create a PR with this and some other fixes.  (it will be interesting to see what the feedback I get)

**Comment by odockal** [comment]
@jeffb-sfdc @mlorinc 
thanks for your proposals, I tend to see the following process like this:
1. Mark former method as deprecated
2. Create new method `getVisibleMarkers`
3. We can release minor version with the change like that and not break actual workflows
4. After some time remove disputed method
5. Implement other suggested updates to the class

**Comment by mlorinc** [comment]
I am looking forward for your PR. After PR I can create follow up tasks from 5th @odockal point as potential future enhancements.

Please note test reliability can be improved by wrapping them in wait statements. For example first test can be rewritten in the following way:

```js
it('Error markers are displayed', async function () {
    // Repeat the following anonymous method for <this.timeout() - 2000> milliseconds. 
    // 2000 is subtracted to make sure Mocha does not hide error. 
    await VSBrowser.instance.driver.wait(async () => {
       // generally, there are 3 marker types (warning, error, and file - file just contains other markers though)
       // we want to see the errors
       const errors = await view.getAllMarkers(MarkerType.Error);

       // assert that there are errors (there should be about 8 in the file)
       return errors.length > 5;
     }, this.timeout() - 2000, 'Could not get more than 5 markers.');
});
```

**Comment by jeffb-sfdc** [comment]
@mlorinc thanks for the tip about using `VSBrowser.instance.driver.wait()`.  I've been adding `await new Promise(resolve => setTimeout((resolve), 3000))` in some places.  Are there any pros/cons with using driver.wait() vs setTimeout()?

**Comment by mlorinc** [comment]
I think the main disadvantage of setTimeout is its blocking nature. Tests are idle for n milliseconds which I think is acceptable for smaller timeouts because lesser code complexity is worth it. I usually use it when I know there will be small time window when element will not be accessible. For example: open file from file explorer and find its editor.

On other hand `driver.wait` calls given function until it times out or truthy value is returned. Delay between attempts can be changed by 4th argument. This approach is better suited for StaleElementReferenceErrors because in this case element must be created again.

**setTimeout pros**

- low code complexity
- easier to debug with debugger tools

**setTimeout cons**

- test execution is throttled by fixed timeouts
- when testing on multiple machines the first con can be more significant and tests are likely to be flaky

**wait pros**

- tests are not blocked
- ideal for solving StaleElementReferenceError
- better when executing tests on multiple machines

**wait cons**

- If there is only one wait the code complexity is quite low. However if there are nested wait calls it can get interesting.
- harder to debug with debugger tools because time out will throw a TimeoutError
- timeout management across multiple wait calls in single test case can be tricky

**Comment by odockal** [comment]
@jeffb-sfdc Please, is there some update on the PR you wanted to work on?

**Comment by jeffb-sfdc** [comment]
OK, I posted https://github.com/redhat-developer/vscode-extension-tester/pull/502
